exercises/hardware_access_control/README.md

```diff
@@ -0,0 +1,160 @@
+Failed to toggle an LED
+```
+
+
+## Part 3
+
+This is great and all, but how do we stop a compartment bypassing `gpio_access` and using `MMIO_CAPABILITY` directly?
+In other words, how do we ensure that only `gpio_access` has access to the GPIOs?
+
+Luckily the linker has all the information needed to check which compartments can access the GPIO MMIOs.
+It outputs this information in a JSON report with the rest of the build artefacts.
+To automate checking this report, we can use `cheriot-audit` which should already be in your path.
+
+`cheriot-audit` allows you to query the JSON report and assert certain rules are followed.
+You do this with a language called [Rego][], but don't worry you won't have to learn it for this exercise.
+There are some pre-written rules in the [`gpio_access.rego`] module.
+Let's first look at `only_gpio_access_has_access`.
+It uses `mmio_allow_list` from `cheriot-audit`'s included compartment package to check only the `gpio_access` compartment has access to the GPIO MMIOs.
+If we run this on the part 2 firmware image's JSON report, it will return true.
+However, when run against the part 1 firmware image's report it will return false, because the `blinky_raw` and `led_walk_raw` are not in the allow list.
+
+[Rego]: https://www.openpolicyagent.org/docs/latest/policy-language/
+[`gpio_access.rego`]: ../../exercises/hardware_access_control/part_3/gpio_access.rego
+
+```sh
+# A convenience function
+check_gpio_access() {
+    cheriot-audit \
+        --board cheriot-rtos/sdk/boards/sonata-prerelease.json \
+        --module exercises/hardware_access_control/part_3/gpio_access.rego \
+        --query "data.gpio_access.$1" \
+        --firmware-report "$2"
+}
+# This should return true
+check_gpio_access only_gpio_access_has_access \
+    build/cheriot/cheriot/release/hardware_access_part_2.json
+# This should return false
+check_gpio_access only_gpio_access_has_access \
+    build/cheriot/cheriot/release/hardware_access_part_1.json
+```
+
+There's a second rule, `whitelisted_compartments_only`, which adds an addition condition that only `led_walk_dynamic` and `blinky_dynamic` can use `gpio_access`.
+We can use this to restrict which compartments have access to the GPIO via `gpio_access`.
+
+```sh
+check_gpio_access whitelisted_compartments_only \
+    build/cheriot/cheriot/release/hardware_access_part_2.json
+```
+
+The above should return true as both compartments are in the allow list.
+Try removing one of the compartments from the allow list given to `compartment_allow_list` in [`gpio_access.rego`][] and check the result of the above command is no longer true.
+
+One can browse the other functions available as part of the `cheriot-audit`'s included compartment package in [`cheriot-audit`'s readme][compartment package].
+
+[compartment package]: https://github.com/CHERIoT-Platform/cheriot-audit/blob/main/README.md#the-compartment-package
+
+## Part ∞
+
+Where to go from here...
```

Comment:
I'd suggest at this point adding a reference to joining the Technical Interest Group - approach one of us for joining that.
Also suggest that there is the Technology Access Programme - TAP - Cohort 6 that lowRISC is helping to adjudicate if people are interested in taking the demos further.
They should also reach out to us via info@lowrisc.org if they have collaboration ideas.

If people have issues they can reach us for the next month on ches-2024@lowrisc.org

Reply:
Addressed in: https://github.com/lowRISC/sonata-software/pull/42